src/mqtt_machine.erl

```diff
@@ -0,0 +1,121 @@
+                   {unregister, client_id(), pid()} |
+                   list.
+
+-spec init(config()) -> state().
+init(_Conf) ->
+    #state{}.
+
+-spec apply(map(), command(), state()) ->
+    {state(), reply(), ra_machine:effects()}.
+apply(Meta, {register, ClientId, Pid}, #state{client_ids = Ids} = State0) ->
+    {Effects, Ids1} =
+        case maps:find(ClientId, Ids) of
+            {ok, OldPid} when Pid =/= OldPid ->
+                Effects0 = [{demonitor, process, OldPid},
+                            {monitor, process, Pid},
+                            {mod_call, gen_server2, cast, [OldPid, duplicate_id]}],
+                {Effects0, maps:remove(ClientId, Ids)};
+            error ->
+              Effects0 = [{monitor, process, Pid}],
+              {Effects0, Ids}
+        end,
+    State = State0#state{client_ids = maps:put(ClientId, Pid, Ids1)},
+    {State, ok, Effects ++ snapshot_effects(Meta, State)};
+
+apply(Meta, {unregister, ClientId, Pid}, #state{client_ids = Ids} = State0) ->
+    State = case maps:find(ClientId, Ids) of
+      {ok, Pid}         -> State0#state{client_ids = maps:remove(ClientId, Ids)};
+      %% don't delete client id that might belong to a newer connection
+      %% that kicked the one with Pid out
+      {ok, _AnotherPid} -> State0;
+      error             -> State0
+    end,
+    {State, ok, [{demonitor, process, Pid}] ++ snapshot_effects(Meta, State)};
+
+apply(Meta, {down, DownPid, _}, #state{client_ids = Ids} = State0) ->
+    Ids1 = maps:filter(fun (_ClientId, Pid) when Pid =:= DownPid ->
+                               false;
+                            (_, _) ->
+                               true
+                       end, Ids),
+    State = State0#state{client_ids = Ids1},
+    Delta = maps:keys(Ids) -- maps:keys(Ids1),
+    Effects = lists:map(fun(Id) ->
+                  [{mod_call, rabbit_log, debug,
+                    ["MQTT connection with client id '~s' failed", [Id]]}] end, Delta),
+    {State, ok, Effects ++ snapshot_effects(Meta, State)};
+
+apply(Meta, {leave, Node}, #state{client_ids = Ids} = State0) ->
+    Ids1 = maps:filter(fun (_ClientId, Pid) -> node(Pid) =/= Node end, Ids),
+    Delta = maps:keys(Ids) -- maps:keys(Ids1),
+
+    Effects = lists:foldl(fun (ClientId, Acc) ->
+                          Pid = maps:get(ClientId, Ids),
+                          [
+                            {demonitor, process, Pid},
+                            {mod_call, gen_server2, cast, [Pid, decommission_node]},
+                            {mod_call, rabbit_log, debug,
+                              ["MQTT will remove client ID '~s' from known "
+                               "as its node has been decommissioned", [ClientId]]}
+                          ]  ++ Acc
+                          end, [], Delta),
+
+    State = State0#state{client_ids = Ids1},
+    {State, ok, Effects ++ snapshot_effects(Meta, State)};
+
+apply(Meta, list, #state{client_ids = Ids} = State) ->
```

Comment:
This doesn't need to be a command. Use either `ra:consistent_query/2` or `ra:leader_query/2` or `ra:local_query/2` to query the state.